llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -1532,43 +1532,58 @@ Expected<uint64_t> AtomRef::materialize(MCCASReader &Reader,
 }
 
 Expected<MCAlignFragmentRef>
-MCAlignFragmentRef::create(MCCASBuilder &MB, const MCAlignFragment &F,
+MCAlignFragmentRef::create(MCCASBuilder &MB, const MCFragment &F,
                            unsigned FragmentSize,
                            ArrayRef<char> FragmentContents) {
   Expected<Builder> B = Builder::startNode(MB.Schema, KindString);
   if (!B)
     return B.takeError();
 
-  uint64_t Count = FragmentSize / F.getFillLen();
-  if (F.hasEmitNops()) {
-    // Write 0 as size and use backend to emit nop.
-    writeVBR8(0, B->Data);
+  writeVBR8(FragmentContents.size(), B->Data);
+  B->Data.append(FragmentContents.begin(), FragmentContents.end());
+  uint64_t Count = (FragmentSize - F.getFixedSize()) / F.getAlignFillLen();
+  if (F.hasAlignEmitNops()) {
+    // Write 1 to signify that it has nops.
+    writeVBR8(1, B->Data);
     if (!MB.Asm.getBackend().writeNopData(MB.FragmentOS, Count,
                                           F.getSubtargetInfo()))
       report_fatal_error("unable to write nop sequence of " + Twine(Count) +
                          " bytes");
     B->Data.append(MB.FragmentData);
     return get(B->build());
   }
+  // Write 0 to signify that it has nops.
+  writeVBR8(0, B->Data);
   writeVBR8(Count, B->Data);
-  writeVBR8(F.getFill(), B->Data);
-  writeVBR8(F.getFillLen(), B->Data);
+  writeVBR8(F.getAlignFill(), B->Data);
+  writeVBR8(F.getAlignFillLen(), B->Data);
   return get(B->build());
 }
 
 Expected<uint64_t> MCAlignFragmentRef::materialize(MCCASReader &Reader,
                                                    raw_ostream *Stream) const {
-  uint64_t Count;
+  uint64_t Count, FragContentSize, HasNops;
   auto Remaining = getData();
   auto Endian = Reader.getEndian();
-  if (auto E = consumeVBR8(Remaining, Count))
+  if (auto E = consumeVBR8(Remaining, FragContentSize))
+    return std::move(E);
+
+  *Stream << Remaining.substr(0, FragContentSize);
+
+  Remaining = Remaining.drop_front(FragContentSize);
```

Comment:
If Remaining is a `StringRef`, maybe `consume_front` is a better API here, to both extract and shrink in one go.